frame/support/src/storage/types/counted_nmap.rs

```diff
@@ -0,0 +1,1446 @@
+use sp_std::prelude::*;
+
+/// A wrapper around a `StorageNMap` and a `StorageValue<Value=u32>` to keep track of how many items
+/// are in a map, without needing to iterate over all of the values.
+///
+/// This storage item has some additional storage read and write overhead when manipulating values
+/// compared to a regular storage map.
+///
+/// For functions where we only add or remove a value, a single storage read is needed to check if
+/// that value already exists. For mutate functions, two storage reads are used to check if the
+/// value existed before and after the mutation.
+///
+/// Whenever the counter needs to be updated, an additional read and write occurs to update that
+/// counter.
+pub struct CountedStorageNMap<
+	Prefix,
+	Key,
+	Value,
+	QueryKind = OptionQuery,
+	OnEmpty = GetDefault,
+	MaxValues = GetDefault,
+>(core::marker::PhantomData<(Prefix, Key, Value, QueryKind, OnEmpty, MaxValues)>);
+
+/// The requirement for an instance of [`CountedStorageNMap`].
+pub trait CountedStorageNMapInstance: StorageInstance {
+	/// The prefix to use for the counter storage value.
+	type CounterPrefix: StorageInstance;
+}
+
+// Private helper trait to access map from counted storage double map
+trait MapWrapper {
+	type Map;
+}
+
+impl<P: CountedStorageNMapInstance, K, V, Q, O, M> MapWrapper
+	for CountedStorageNMap<P, K, V, Q, O, M>
+{
+	type Map = StorageNMap<P, K, V, Q, O, M>;
+}
+
+type CounterFor<P> =
+	StorageValue<<P as CountedStorageNMapInstance>::CounterPrefix, u32, ValueQuery>;
+
+/// On removal logic for updating counter while draining upon some prefix with
+/// [`crate::storage::PrefixIterator`].
+pub struct OnRemovalCounterUpdate<Prefix>(core::marker::PhantomData<Prefix>);
+
+impl<Prefix: CountedStorageNMapInstance> crate::storage::PrefixIteratorOnRemoval
+	for OnRemovalCounterUpdate<Prefix>
+{
+	fn on_removal(_key: &[u8], _value: &[u8]) {
+		CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+	}
+}
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
+	crate::storage::generator::StorageNMap<Key, Value>
+	for CountedStorageNMap<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
+where
+	Prefix: CountedStorageNMapInstance,
+	Key: super::key::KeyGenerator,
+	Value: FullCodec,
+	QueryKind: QueryKindTrait<Value, OnEmpty>,
+	OnEmpty: Get<QueryKind::Query> + 'static,
+	MaxValues: Get<Option<u32>>,
+{
+	type Query = QueryKind::Query;
+	fn module_prefix() -> &'static [u8] {
+		Prefix::pallet_prefix().as_bytes()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		Prefix::STORAGE_PREFIX.as_bytes()
+	}
+	fn from_optional_value_to_query(v: Option<Value>) -> Self::Query {
+		QueryKind::from_optional_value_to_query(v)
+	}
+	fn from_query_to_optional_value(v: Self::Query) -> Option<Value> {
+		QueryKind::from_query_to_optional_value(v)
+	}
+}
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues> crate::storage::StoragePrefixedMap<Value>
+	for CountedStorageNMap<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
```

Comment:
This is not correct, StoragePrefixedMap is a trait which consider that all values of the map are after one prefix. but also doesn't account for anything like the counter.

Maybe we should improve documentation of StoragePrefixedMap with something like this:
```diff
diff --git a/frame/support/src/storage/mod.rs b/frame/support/src/storage/mod.rs
index 6a88eb676ea..c0e44043bae 100644
--- a/frame/support/src/storage/mod.rs
+++ b/frame/support/src/storage/mod.rs
@@ -1178,6 +1178,13 @@ impl<T> Iterator for ChildTriePrefixIterator<T> {
 /// ```nocompile
 /// Twox128(module_prefix) ++ Twox128(storage_prefix)
 /// ```
+///
+/// # Warning
+///
+/// The trait methods default implementations can result in inconsistent operation if some
+/// information on the keys or values are stored outside the map.
+/// For instance a counter counting for values would get inconsistent if [`Self::clear`] default
+/// implementations is called.
 pub trait StoragePrefixedMap<Value: FullCodec> {
        /// Module prefix. Used for generating final key.
        fn module_prefix() -> &'static [u8];
```

IMO just don't implement this trait. similarly to CountedStorageMap.

Reply:
fixed, thanks